### ISSUE TITLE
Feature Request: Automatically generate $pvc-luks-key secret for encrypted PVCs

### ISSUE THREAD

**Issue by davidgubler** [cloudscale-ch/csi-cloudscale#55]
### What did you do?

What I would like to do:
Create a PVC with storage class "cloudscale-volume-ssd-luks" WITHOUT creating the corresponding secret $pvc-luks-key.

### What did you expect to happen?

What I would like to happen:
The Cloudscale CSI driver generates a luks encryption secret of suitable length, puts it into a secret $pvc-luks-key, and then sets up the encrypted volume as it does now. The automatically generated secret $pvc-luks-key gets the same labels as the $pvc (as far as possible).

### Configuration:

Since this is a feature request there are no logs or configuration information.

* CSI Version: 3.5.0 at the time of writing this feature request

* Kubernetes Version: 1.24

* Cloud provider/framework version: OpenShift 4.11.28

### Reasoning

TL;DR: We want to use encrypted volumes without having to know the implementation details of csi-cloudscale.

General reasons:
* Creating the secret manually does not provide any value to the user of an encrypted volume. Getting rid of this manual step improves the user experience and simplifies tools which want to automatically set up encrypted volumes.
* The user can still access the generated secret if he desires to. As far as I understand this the secret is pretty much useless to the user anyway because the user cannot access the underlying encrypted block device. (not 100% sure about this)
* Security is improved by ensuring that the secret has the maximum possible entropy (instead of potentially reduced entropy due to insecure secret generation methods)
* This change is backwards compatible to how things work now simply by only generating the secret if it doesn't already exist. If it already exists use the existing secret.

Specific reasons in our case:
* We're working on tooling to abstract away the complexity of allocating PVs from the user as far as possible (https://github.com/vshn/k8ify ). It would be very helpful if we could use encrypted volumes just by specifying the storage class. With the current implementation this isn't possible; instead the abstraction layer (k8ify) has to have information about the target cluster, in particular it needs to know that the storage driver is csi-cloudscale and it has to know about the luks secrets. This is essentially preventing us from implementing support for encrypted volumes for now, because the step that generates K8s manifests (a build pipeline running k8ify) has little or no information about the target K8s cluster.
* The secret needs to copy the labels from the PVC because that allows us to automatically delete the secret when the PVC gets deleted. Without the labels we would again need to know specifics of csi-cloudscale (in particular how the secret name is constructed).

Thanks for your consideration!

**Comment by alakae** [comment]
Hi @davidgubler 

Thank you for raising this issue. My main concern is backward compatibility. We'd prefer a flag that explicitly enables auto-generating the LUKS key over just making this the new default. However, to my knowledge it is not trivial to add new parameters to existing `StorageClass` objects. I'll have to investigate this further and discuss it with the team.

I have added this request to our internal backlog. Feel free to contact us at any time if you have any questions or concerns.

Best regards,
Alain

**Comment by davidgubler** [comment]
Hi @alakae,

Thanks for the feedback!

In my opinion backwards compatibility should not be a problem because I would expect the implementation to auto-generate the LUKS key only if no secret with the name exists. With this implementation users could continue to set up encrypted volumes as they do now.

Best regards,

David

**Comment by alakae** [comment]
Hi @davidgubler 

In the meantime I had an opportunity to have an in-depth look at your 
feature request. 

Before I go into the details, let me comment on this:

> As far as I understand this the secret is pretty much useless to the user 
> anyway because the user cannot access the underlying encrypted block device. 
> (not 100% sure about this)

You can access the LUKS partition that the cloudscale.ch CSI driver creates
on the volume on any virtual machine (even outside the clsuter) by running
`cryptsetup luksOpen` and  providing the secret. We therefore think it's essential
for disaster recovery to have access to the secret even if the cluster is unavailable.

Now onto the actual question of lazily creating the LUKS secret if it is not provided
by the user. The LUKS secret is configured as `StorageClass Secret` of type
`Node Stage Secret` (see [here](https://github.com/cloudscale-ch/csi-cloudscale/blob/1d6340d2b48d63b203e1156fa5f07074c318dfae/deploy/kubernetes/releases/csi-cloudscale-v3.5.1.yaml#L46)).
`NodeStageVolume` is the CSI Operation in which the volume is attached to a server
and the LUKS partition is opened.

I'm sure you're aware of the fact, that CSI drivers typically are deployed
together with a number of sidecar containers, one of them being the
[external-provisioner](https://github.com/kubernetes-csi/external-provisioner).
It is this sidecar that is responsible for retrieving the `StorageClass Secrets`
from Kubernetes and passing them to the cloudscale CSI driver implementation (
in a gRPC Call). This is explained in [here](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#node-stage-secret)
in detail for `Node Stage Secrets`. I find the following two paragraphs especially
interesting:

> If no such secret exists in the Kubernetes API, or the kubelet is unable to fetch it, the mount device operation fails.
> 
> If the secret is retrieved successfully, the kubelet passes it to the CSI driver in the NodeStageVolumeRequest.secrets field.

This means that `NodeStageVolume` is not called if the secret is absent, and it must
exist before the `external-provisioner` tries to access the secret.

This has led me to the idea of making a secret in the `CreateVolume` call, which happens
before `NodeStageVolume`. But I think this not a good idea for two reasons:
 - The LUKS secret is a `Node Stage Secret` and should not be touched in the `CreateVolume`.
 - `CreateVolume` is a gRPC call which gets _passed_ the `Create/Delete Volume Secret`.
   There is not really an "official" API for _creating_ a secret. While it might 
   be technically possible to communicate with `kube-api` directly to query and create arbitrary secrets
   by creating  a `Clientset`, this looks like a bad practice to me and not how
   `Storage Class Secrets` are intended to be used.

Another options would of course be to fork `external-provisioner`. But again,
if this is required, it is a clear sign to me, that we are not using the CSI ecosystem
as intended.

In summary, while I can clearly understand your use-case, I haven't found a clean
way of implementing it. If you find any documentation that describes your use-case or 
if you are aware of any other CSI driver that implements this pattern, I am more than
happy to have a look at that.

In the meantime, please understand that we cannot help you right now, at least until
we have found a "clean" way of implementing this.

I am mentioning @ferdinandhuebner here, the original author auf the LUKS feature in our CSI 
driver. Maybe he stops by and has a clever idea ;-).

**Comment by ferdinandhuebner** [comment]
Hi @alakae and @davidgubler 

I agree with @alakae that this is not something a CSI driver should be responsible for. 

While I didn't follow the evolution of the CSI spec since the last time this feature was requested (see #10), I believe that the fundamentals haven't changed. Please take a look at my comment in https://github.com/cloudscale-ch/csi-cloudscale/issues/10#issuecomment-551036944 for an explanation.

A [dynamic admission controller](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) could fill the gap. Please note that this would leave generated LUKS secrets lying around if the volumes are deleted. One could implement some kind of cleanup job in the admission controller though.

Another approach would be to implement a custom resource definition (CRD) with a controller that handles that custom resource. Instead of deploying a persistent volume claim, you'd deploy an instance of your custom resource. The controller would then provision a LUKS secret and a persistent volume claim. 
As far as I understand @davidgubler, this wouldn't solve their problem because `k8ify` would need to know if it's a cloudscale.ch cluster and use the CRD instead of PVCs.

On a sidenote: It seems to me, that you might not even need LUKS-encrypted volumes in the first place. Your data is encrypted 'at rest' anyway (see https://www.cloudscale.ch/en/news/2019/07/25/bluestore-encryption-and-nvme-only-storage). 
We implemented LUKS-encryption at the time, because it wasn't clear to us if data was encrypted at rest and we wanted support for "customer managed encryption keys" for a security-conscious client. If you don't care about that and just need encrypted data at rest, you don't need to use the LUKS feature of the CSI driver.

**Comment by alakae** [comment]
@ferdinandhuebner Thanks a lot for your explanation and opinion.

**Comment by davidgubler** [comment]
Thanks for the feedback and detailed explanations.

After reassessing the situation I've come to the conclusion that we will have to implement CSI driver-specific mechanisms to enable encryption anyway, because every CSI driver seems to have wildly different ideas on how encryption needs to be set up, and thus a good, universal design is not possible with the current sorry state of the CSI ecosystem. I guess encryption would need to be standardized on the CSI level first before the situation can improve.

Therefore I'm closing this.
